azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/message_queue.py

```diff
@@ -0,0 +1,142 @@
+
+    def _input_message_queue_on_disable(self):
+        # Disable input messages as a feature if all input queues are disabled
+        if all(not queue.enabled for queue in self.input_message_queues.values()):
+            self._disable_feature("input")
+
+    def _c2d_message_queue_on_disable(self):
+        self._disable_feature("c2d")
+
+    def _enable_feature(self, feature_name):
+        """
+        To enable a specific feature on the internal client.Some of the features that can be enabled
+        are "receiving input messages" and "receiving cloud to device messages"
+        :param feature_name: The specific feature which needs to be enabled. Feature names can be:-
+        "input" , "c2d" etc.
+        "inputs" and on the attribute "on_c2d_message" for "c2d"
+        :raises ValueError if feature_name is not amongst "input" or "c2d"
+        """
+        enable_complete = Event()
+
+        def callback():
+            enable_complete.set()
+
+        if feature_name == "input":
+            self._transport.enable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.enable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        enable_complete.wait()
+
+    def _disable_feature(self, feature_name):
+        disable_complete = Event()
+
+        def callback():
+            disable_complete.set()
+
+        if feature_name == "input":
+            self._transport.disable_input_messages(callback)
+        elif feature_name == "c2d":
+            self._transport.disable_c2d_messages(callback)
+        else:
+            logger.error("Feature name has not been defined.Feature names can be 'input' or 'c2d'")
+            raise ValueError("Feature names can be only among 'input' or 'c2d'")
+
+        disable_complete.wait()
+
+    def get_input_message_queue(self, input_name):
+        try:
+            queue = self.input_message_queues[input_name]
+        except KeyError:
+            # Create new MessageQueue for input if it does not yet exist
+            queue = MessageQueue(
+                self._input_message_queue_on_enable, self._input_message_queue_on_disable
+            )
+            self.input_message_queues[input_name] = queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def get_c2d_message_queue(self):
+        if self.c2d_message_queue is None:
+            queue = MessageQueue(
+                self._c2d_message_queue_on_enable, self._c2d_message_queue_on_disable
+            )
+            self.c2d_message_queue = queue
+        else:
+            queue = self.c2d_message_queue
+
+        if queue.enabled is not True:
+            queue.enable()
+        return queue
+
+    def receive_input_message(self, input_name, message_received):
```

Comment:
>receive_input_message [](start = 8, length = 21)

don't worry too much about this, but the pattern for this function (and get_input_message_queue) will be repeated at least twice more (for twin updates and for method calls).  Both of those are like input messages in that we'll have multiple queues for a single feature.